### PR TITLE
Port REACT_MODULE() from react-native-windows to react-native-macos (2/4)

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
# Summary
- Make `REACT_MODULE` usable for macOS (co-op with https://github.com/microsoft/react-native-macos/pull/599 ).
- In this change, comments are added to involved files to tell developers that, they need to take care about macOS from now on.
- In order to do this, react-native-macos needs to use some files from react-native-windows.
- For development convenience, these files are duplicated in react-native-macos.
- After the porting is done, I will make another change to eliminate the duplication. Files will be shared from one single source.

# Notice
This change also fixes a few issues:
- Workaround XCode compiler issue
  - optional<T>::value() is missing (it is added in macOS 10.14, but it is required to support 10.13)
  - hash<pair<T, U>> is missing (same reason)
  - lambda cannot capture objects that only have rvalue copy ctor (this should be xcode's bug)
- Function returning `ReactPromise` is not currectly handled when `Reject` is called.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6023)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/JsiReader.h [hunk 1]
```diff
@@ -8,6 +11,15 @@
 
 namespace winrt::Microsoft::ReactNative {
 
+// workaround: in macOS 10.13, optional<T>::value is not available
+
+#ifndef __APPLE__
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/JsiReader.h:16]: >__APPLE__ [](start = 8, length = 9)

What is the Apple platform alternative?

File: vnext/Microsoft.ReactNative/JsiReader.h [hunk 2]
```diff
@@ -8,6 +11,15 @@
...
+template <typename T>
+decltype(auto) ReadOptional(std::optional<T> &opt) {
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/JsiReader.h:18]: >decltype(auto) [](start = 0, length = 14)

It could be just 'auto'

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Microsoft.ReactNative/JsiReader.h (hunk 1 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/JsiReader.h:16]: It calls `__get` instead of `value`

File: vnext/Microsoft.ReactNative/JsiReader.h (hunk 2 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/JsiReader.h:18]: Here I want to keep the same exact reference type, so I am intended to use `decltype(auto)`

**Pull request merged**